### PR TITLE
New functions in Land Tiles, Gumps, Texture and Tiledata tabs

### PR DESCRIPTION
**Pull request by Maybacco**
New functions: 
- Added some "Insert starting from" buttons in the Land Tiles, Gumps and Textures tabs. It allows user to add multiple files at once starting from chosen index.
- Added "Set Textures" option in Tiledata tab to set every landtile to its textureId.

### CODE REVIEW

File: UoFiddler.Controls/UserControls/GumpControl.cs [hunk 1]
```diff
@@ -697,5 +697,104 @@ private void Gump_KeyUp(object sender, KeyEventArgs e)
...
+                dialog.Title = $"Choose image file to insert at 0x{index:X}";
+                dialog.CheckFileExists = true;
+                dialog.Filter = "Image files (*.tif;*.tiff;*.bmp)|*.tif;*.tiff;*.bmp";
+                if (dialog.ShowDialog() != DialogResult.OK)
+                {
+                    return;
+                }
+
+
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/GumpControl.cs:727]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/GumpControl.cs [hunk 2]
```diff
@@ -697,5 +697,104 @@ private void Gump_KeyUp(object sender, KeyEventArgs e)
...
+            {
+                return;
+            }
+
+            if (!Utils.ConvertStringToInt(InsertStartingFromTb.Text, out int index, 0, Gumps.GetCount()))
+            {
+                return;
+            }
+
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/GumpControl.cs:713]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/GumpControl.cs [hunk 3]
```diff
@@ -697,5 +697,104 @@ private void Gump_KeyUp(object sender, KeyEventArgs e)
...
+                    for(int i = 0; i < dialog.FileNames.Length; i++)
+                    {
+                        var currentIdx = index + i;
+                        AddSingleGump(dialog.FileNames[i], currentIdx);
+                    }
+                }
+
+                
+
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/GumpControl.cs:739]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/LandTilesControl.cs [hunk 4]
```diff
@@ -869,5 +869,117 @@ private void LandTiles_KeyUp(object sender, KeyEventArgs e)
...
+                {
+                    for(int i = 0; i < dialog.FileNames.Length; i++)
+                    {
+                        var currentIdx = index + i;
+                        AddSingleLandTile(dialog.FileNames[i], currentIdx);
+                    }
+                }
+
+
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/LandTilesControl.cs:908]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/LandTilesControl.cs [hunk 5]
```diff
@@ -869,5 +869,117 @@ private void LandTiles_KeyUp(object sender, KeyEventArgs e)
...
+                listView1.View = View.Tile;
+
+                if (listView1.SelectedItems.Count == 1)
+                {
+                    listView1.SelectedItems[0].Selected = false;
+                }
+
+                //item.Selected = true;
+                //item.Focused = true;
+                //item.EnsureVisible();
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/LandTilesControl.cs:919]: If this code is not needed then please remove it.

File: UoFiddler.Controls/UserControls/LandTilesControl.cs [hunk 6]
```diff
@@ -869,5 +869,117 @@ private void LandTiles_KeyUp(object sender, KeyEventArgs e)
...
+        private bool CheckForIndexes(int baseIndex, int count)
+        {
+
+            for(int i = baseIndex; i < baseIndex + count; i++)
+            {
+                if (Art.IsValidLand(i))
+                {
+                    return false;
+                }
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/LandTilesControl.cs:932]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/LandTilesControl.cs [hunk 7]
```diff
@@ -869,5 +869,117 @@ private void LandTiles_KeyUp(object sender, KeyEventArgs e)
...
+                //item.Selected = true;
+                //item.Focused = true;
+                //item.EnsureVisible();
+            }
+        }
+
+        private bool CheckForIndexes(int baseIndex, int count)
+        {
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/LandTilesControl.cs:925]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/TextureControl.cs [hunk 8]
```diff
@@ -546,5 +546,105 @@ private void ExportAllTextures(ImageFormat imageFormat)
...
+
+            if (!Utils.ConvertStringToInt(StartingPosTb.Text, out int index, 0, Textures.GetIdxLength()))
+            {
+                return;
+            }
+
+            //if (Textures.TestTexture(index))
+            //{
+            //    return;
+            //}
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/TextureControl.cs:565]: If this code is not needed then please remove it.

File: UoFiddler.Controls/UserControls/TextureControl.cs [hunk 9]
```diff
@@ -546,5 +546,105 @@ private void ExportAllTextures(ImageFormat imageFormat)
...
+
+                if (CheckForIndexes(index, fileCount))
+                {
+                    for (int i = 0; i < fileCount; i++)
+                    {
+                        AddSingleTexture(dialog.FileNames[i], index + i);
+                    }
+                }
+
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/TextureControl.cs:589]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/TextureControl.cs [hunk 10]
```diff
@@ -546,5 +546,105 @@ private void ExportAllTextures(ImageFormat imageFormat)
...
+                listView1.View = View.Details; // that works fascinating
+                listView1.View = View.Tile;
+
+                if (listView1.SelectedItems.Count == 1)
+                {
+                    listView1.SelectedItems[0].Selected = false;
+                }
+                Options.ChangedUltimaClass["Texture"] = true;
+
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/TextureControl.cs:598]: Please remove redundant empty lines.

File: UoFiddler.Controls/UserControls/TileDataControl.cs [hunk 11]
```diff
@@ -1634,5 +1634,22 @@ private void ItemsContextMenuStrip_Opening(object sender, System.ComponentModel.
...
             }
         }
+
+        private void textBoxTexID_DoubleClick(object sender, EventArgs e)
+        {
+            int index = (int)treeViewLand.SelectedNode.Tag;
+            textBoxTexID.Text = $"{index}";
+        }
+
+        private void setTexturesToolStripMenuItem_Click(object sender, EventArgs e)
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/TileDataControl.cs:1644]: Method names should start with capital letter -> SetTexturesToolStripMenuItem_Click

File: UoFiddler.Controls/UserControls/TileDataControl.cs [hunk 12]
```diff
@@ -1634,5 +1634,22 @@ private void ItemsContextMenuStrip_Opening(object sender, System.ComponentModel.
                 }
             }
         }
+
+        private void textBoxTexID_DoubleClick(object sender, EventArgs e)
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/TileDataControl.cs:1638]: Method names should start with capital letter -> TextBoxTexID_DoubleClick

File: UoFiddler.Controls/UserControls/LandTilesControl.cs [hunk 13]
```diff
@@ -869,5 +869,117 @@ private void LandTiles_KeyUp(object sender, KeyEventArgs e)
             e.SuppressKeyPress = true;
             e.Handled = true;
         }
+
+        private void toolStripTextBox1_KeyDown(object sender, KeyEventArgs e)
```
Comment (AsYlum-) [inline UoFiddler.Controls/UserControls/LandTilesControl.cs:873]: Method names should start with capital letter -> ToolStripTextBox1_KeyDown

**Review by AsYlum- — CHANGES_REQUESTED** [review]
Thank you for the PR. Overall looks good but there are few things to clean up before merging. 

For LandTilesControl and TextureControl we have alternative ones LandTilesAlternativeControl and TextureAlternativeControl can you please add the features there as well? Alternative controls uses new faster TileView control and will be main and only controls soon.

**Review by Maybacco — COMMENTED** [review]
Added the feature in LandTilesAlternative and TextureAlternative controls.

Fixed the blank lines and naming.

Added some comments to describe the utility functions.

Got  something strange on TextureAlternative, why are we doing the & with FFF and not with 3FFF as in LandTiles? Are those two files different in size?

**Comment by AsYlum-** [comment]
It's probably a bug :) looks like we can assume that there is 1:1 texture landtile ratio and it should be 0x3FFF right now

**Comment by AsYlum-** [comment]
Would you like to do that update with 0x3FFF or should I take over after merge? :)

**Comment by AsYlum-** [comment]
I've edited PR title and description to better reflect what is changed.

**Review by AsYlum- — COMMENTED** [review]
This looks like it may fail.

**Comment by Maybacco** [comment]
It actually gives a "Index out of bounds" on the matrix. I'll look into it to crash-fix that and i'll submit the fix asap!

File: UoFiddler.Controls/UserControls/TextureAlternativeControl.cs [hunk 14]
```diff
@@ -503,5 +504,109 @@ private void ExportAllTextures(ImageFormat imageFormat)
...
+            }
+        }
+
+        /// <summary>
+        /// Check if all the indexes from baseIndex to baseIndex + count are valid
+        /// </summary>
+        /// <param name="baseIndex">Starting Index</param>
+        /// <param name="count">Number of the indexes to check.</param>
+        /// <returns></returns>
+        private bool CheckForIndexes(int baseIndex, int count)
```
[the remark this answers is not in the document]
Comment (Maybacco) [reply UoFiddler.Controls/UserControls/TextureAlternativeControl.cs:557]: I was also trying with the "Insert at.." menu item and by simply writing 0x4000 the software crashes with a Index out of bound.

I'll try to check the TestTexture function to return a false with a out of bound.

[the remark this answers is not in the document]
Comment (AsYlum-) [reply UoFiddler.Controls/UserControls/TextureAlternativeControl.cs:557]: It should fail at 0x4000 as last valid index is 0x3FFF. 0x4000 is array size that holds items and it's 0-based so 0-0x3FFF :) In CheckForIndexes() method you probably need update if() to also check if current index is not outside of maximum index.

`if (Textures.TestTexture(i) && i < Textures.GetIdxLength() )`

And similar change probably needs to be done for Gumps, Land Tiles and regular Textures control.

**Review by AsYlum- — APPROVED** [review]

**Comment by AsYlum-** [comment]
Do you plan to add anything else here? If not then I will merge this and put updated pre-release for further testing.

**Comment by Maybacco** [comment]
For now i don't think so, i'll probably add the same feature in the Items tab. If you want i can add that in the next days and i'll update on this PR

**Comment by AsYlum-** [comment]
I think it would be better to create new PR for Items changes. 

One tip: It's better to create new feature branch in your fork based on master than commit to forked master and creating PR out of it. Because when I merge this PR right now and you will try to update your fork you will probably get some issues. Better approach is to keep main branches clean and work on feature branches created out of them :)

And once again thank you for your work.

**Pull request merged**